File: packages/telegram_bot/src/telegram_bot/spam.py
import re

MINIMUM_MESSAGE_LENGTH = 3
MAXIMUM_MESSAGE_LENGTH = 250
MAX_EMOJIS = 5

_EMOJI_PATTERN = re.compile(
    "[\U0001f600-\U0001f64f]+",
    flags=re.UNICODE,
)


def is_spam(text: str) -> bool:
    if len(text) < MINIMUM_MESSAGE_LENGTH:
        return True

    if "?" in text:
        return True

    if len(text) > MAXIMUM_MESSAGE_LENGTH:
        return True

    if "http" in text:
        return True

    emojis = [char for match in _EMOJI_PATTERN.findall(text) for char in match]
    return len(emojis) > MAX_EMOJIS

File: packages/telegram-worker/src/spam.ts
export const MINIMUM_MESSAGE_LENGTH = 3
export const MAXIMUM_MESSAGE_LENGTH = 250
export const MAX_EMOJIS = 5

// Emoji range: Emoticons block, U+1F600–U+1F64F.
const EMOJI_PATTERN = /[\u{1f600}-\u{1f64f}]/gu

export function isSpam(text: string): boolean {
    if (text.length < MINIMUM_MESSAGE_LENGTH) {
        return true
    }
    if (text.includes('?')) {
        return true
    }
    if (text.length > MAXIMUM_MESSAGE_LENGTH) {
        return true
    }
    if (text.includes('http')) {
        return true
    }
    const emojis = text.match(EMOJI_PATTERN)
    return (emojis?.length ?? 0) > MAX_EMOJIS
}
